pkg/services/auth/internal/gitlab.go

```diff
@@ -0,0 +1,66 @@
+package internal
+
+import (
+	"net/url"
+	"strings"
+)
+
+const (
+	gitlabClientIdKey    = "client_id"
+	gitlabRedirectUriKey = "redirect_uri"
+	gitLabCodeKey        = "code"
+	gitlabScheme         = "https"
+	gitlabHost           = "gitlab.com"
+	gitlabClientId       = "a28c1cb975f2cffd749bc7e32a5a1ebce8a3e7a2812ff27b54d2265ea0de3633"
+	gitlabClientSecret   = "6f173cc1018f7e6cfdd76b1912dfd03b15aaf8a4cb3a5aad4b30a219170408eb"
+
+	GitlabVerifierMin    = 43
+	GitlabVerifierMax    = 128
+	GitlabRedirectUriCLI = "http://127.0.0.1:9999/oauth/gitlab/callback"
+	GitlabCallbackPath   = "/oauth/gitlab/callback"
+	GitlabTempServerPort = ":9999"
+)
+
+type GitlabTokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    int64  `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+	CreatedAt    int64  `json:"created_at"`
+}
+
+func GitlabAuthorizeUrl(redirectUri string, scopes []string, verifier CodeVerifier) url.URL {
+	u := url.URL{}
+	u.Scheme = gitlabScheme
+	u.Host = gitlabHost
+	u.Path = "/oauth/authorize"
+
+	params := u.Query()
+	params.Set(gitlabClientIdKey, gitlabClientId)
+	params.Set(gitlabRedirectUriKey, redirectUri)
+	params.Set("response_type", gitLabCodeKey)
+	params.Set("code_challenge", verifier.CodeChallenge())
+	params.Set("code_challenge_method", "S256")
```

Comment:
Honestly, the ones that are const were used more than once.  I can go ahead and switch them back to strings.  I almost didn't do it in the first place because I thought the difference looked weird too.

Reply:
Done